<a href="https://colab.research.google.com/github/NadezhdaGruzdeva/Statistics-Probability/blob/main/Statistics%26Probability_HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import math

1) Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.


In [ ]:
salaries = [100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150] 
salaries

[100,
 80,
 75,
 77,
 89,
 33,
 45,
 25,
 65,
 17,
 30,
 24,
 57,
 55,
 70,
 75,
 65,
 84,
 90,
 150]

In [ ]:
salaries.sort()
salaries

[17,
 24,
 25,
 30,
 33,
 45,
 55,
 57,
 65,
 65,
 70,
 75,
 75,
 77,
 80,
 84,
 89,
 90,
 100,
 150]

In [ ]:
def Average(myList):
   return sum(myList)/len(myList)

In [ ]:
averageSalary = Average(salaries)
averageSalary

65.3

In [ ]:
np.mean(salaries)

65.3

Cредняяя арифметическая з/п = 65,3

In [ ]:
def Variance(myList, ddof):
  average = Average(myList)
  counter = 0 
  myListSize = len(myList)
  for i in range(myListSize):
    counter += (myList[i] - average) ** 2
  return counter/(myListSize - ddof)

In [ ]:
Variance(salaries,0)

950.11

In [ ]:
Variance(salaries,1)

1000.1157894736842

In [ ]:
np.var(salaries,ddof = 0)

950.1099999999999

In [ ]:
np.var(salaries,ddof = 1)

1000.115789473684

Смещенная дисперсия = 950.1, несмещенная - 1000.1

In [ ]:
def StDeviation(myList, ddof):
  return Variance(myList, ddof) ** (1/2)

In [ ]:
StDeviation(salaries,0)

30.823854398825596

In [ ]:
StDeviation(salaries,1)

31.624607341019814

In [ ]:
np.std(salaries,ddof = 0)

30.823854398825592

In [ ]:
np.std(salaries,ddof = 1)

31.62460734101981

Смещенное среднее квадратическое отклонение = 30.8, несмещенное - 31.6

2) В первом ящике находится 8 мячей, из которых 5 - белые. Во втором ящике - 12 мячей, из которых 5 белых. Из первого ящика вытаскивают случайным образом два мяча, из второго - 4. Какова вероятность того, что 3 мяча белые?

In [ ]:
def Combination(k,n): 
  '''Функция для сочетний
  k - количество элементов в наборе
  n - общее число элементов
  '''
  return int(math.factorial(n) / (math.factorial(k) * math.factorial(n - k)))

In [ ]:
def Bernully(k,n,p):
  '''P(X=k) = C_n ^k    *〖p 〗^k*q^(n-k)
позволяющая находить вероятность появления события 
A определённое количество раз при любом числе независимых испытаний
	k - число наступления события (дискретная величина из отрезка [0, n])
	n -  число испытаний 
	p - вероятность наступления события A в  независимых испытаниях, 
	q = 1 – p. Противоположная вероятность

  '''
  return Combination(k,n) * p ** k * (1 - p) ** (n - k)

In [ ]:
P = Bernully(0,2,5/8) * Bernully(3,4,5/12) + Bernully(1,2,5/8) * Bernully(2,4,5/12) + Bernully(2,2,5/8) * Bernully(1,4,5/12)
P

0.3191159095293209



```
# Выбран кодовый формат
```

Ответ: 32% - это неправильное решение. " нельзя использовать формулу Бернулли, потому что вероятность извлечения мяча не постоянная, меняется общее количество мячей и белых мячей
Формула Бернулли используется для повторных независимых испытаний"

Исправленное решение с использованием сочетаний ниже

In [ ]:
allCombBasket1 = Combination(2,8) # все сочетания вынимая 2 мяча из 8 в 1-ой корзне
allCombBasket1

28

In [ ]:
White2pcCombBasket1 = Combination(2,5) # все сочетания вынимая 2 белых мяча из 5 в 1-ой корзне
White2pcCombBasket1

10

In [ ]:
Black2pcCombBasket1 = Combination(2,3) # все сочетания вынимая 2 черных мяча из 3 в 1-ой корзне, то есть вынуть 0 белых
Black2pcCombBasket1

3

In [ ]:
White1pcCombBasket1 = allCombBasket1 - White2pcCombBasket1 - Black2pcCombBasket1   # все сочетания вынимая 1 белый мяч и 1 черный из 1-ой корзины
White1pcCombBasket1

15

Перейдем к сочетаниям 2-ой корзины

In [ ]:
allCombBasket2 = Combination(4,12) # все сочетания вынимая 4 мяча из 12 во 2-ой корзне
allCombBasket2

495

In [ ]:
White4pcCombBasket2 = Combination(4,5) # все сочетания вынимая 4 белых мяча из 5 во 2-ой корзне
White4pcCombBasket2

5

In [ ]:
Black4pcCombBasket2 = Combination(4,7) # все сочетания вынимая 4 черных мяча из 7 во 2-ой корзне, то есть вынуть 0 белых
Black4pcCombBasket2

35

In [ ]:
White2pcBlack2pcCombBasket2 = Combination(2,5) * Combination(2,7) # все сочетания вынимая 2 белых мяча из 5 и 2 черных мяча из 7 во 2-ой корзне
# умножение, так как абстрактно представляется именно так, в конце проверить
White2pcBlack2pcCombBasket2

210

In [ ]:
White1pcBlack3pcCombBasket2 = Combination(1,5) * Combination(3,7) # все сочетания вынимая 1 белых мяча из 5 и 3 черных мяча из 7 во 2-ой корзне
White1pcBlack3pcCombBasket2

175

In [ ]:
White3pcBlack1pcCombBasket2 = Combination(3,5) * Combination(1,7) # все сочетания вынимая 3 белых мяча из 5 и 1 черных мяча из 7 во 2-ой корзне
White3pcBlack1pcCombBasket2

70

In [ ]:
#проверка - ок
allCombBasket2 == (White4pcCombBasket2 + Black4pcCombBasket2 + White2pcBlack2pcCombBasket2 + White1pcBlack3pcCombBasket2 + White3pcBlack1pcCombBasket2)

True

In [ ]:
#расчитаем вероятность
P = (White2pcCombBasket1/allCombBasket1 * White1pcBlack3pcCombBasket2/allCombBasket2 + 
     White1pcCombBasket1/allCombBasket1 * White2pcBlack2pcCombBasket2/allCombBasket2 + 
     Black2pcCombBasket1/allCombBasket1 * White3pcBlack1pcCombBasket2/allCombBasket2)
P

0.3686868686868687

Ответ: 36,9%

3) На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6. Найти вероятность того, что выстрел произведен: a). первым спортсменом б). вторым спортсменом в). третьим спортсменом.

In [ ]:
P1 = Bernully(1,1,0.9) * Bernully(0,1,0.8) * Bernully(0,1,0.6)
P1

0.072

In [ ]:
P2 = Bernully(0,1,0.9) * Bernully(1,1,0.8) * Bernully(0,1,0.6)
P2

0.031999999999999994

In [ ]:
P3 = Bernully(0,1,0.9) * Bernully(0,1,0.8) * Bernully(1,1,0.6)
P3

0.011999999999999993

In [ ]:
P = P1 + P2 + P3
P

0.11599999999999998

Подведем общий итог к 100%, и расчитаем вероятность попадания того или иного стрелка пропорционально полученным выше вероятностям


In [ ]:
P_1 = P1/P
P_1

0.6206896551724138

In [ ]:
P_2 = P2/P
P_2

0.27586206896551724

In [ ]:
P_3 = P3/P
P_3

0.10344827586206892

Ответ: вероятность того, что выстрел произведен: a). первым спортсменом = 62% б). вторым спортсменом = 28%  в). третьим спортсменом = 10%. (Обязательно посмотреть разбор этой задачи потом) - НО СТРЕЛЯЕТ 1 СЧТРЕЛОК, ТАК ЧТО НОВОЕ РЕШЕНИЕ НИЖЕ

Используем теорему Байеса: P(A|B) = P(B|A) * P(A)/P(B)

In [ ]:
# Полноя вероятность события B что попадаеет стрелок в цель
PB = 1/3*0.9 + 1/3 * 0.8 + 1/3*0.6 
PB

0.7666666666666666

In [ ]:
# 3.1) Стреляет и попадает первый стрелок
P1 = 0.9 * 1/3 / PB
P1

0.391304347826087

In [ ]:
# 3.2) Стреляет и попадает 2-ой стрелок
P2 = 0.8 * 1/3 / PB
P2

0.3478260869565218

In [ ]:
# 3.3) Стреляет и попадает 3-ий стрелок
P3 = 0.6 * 1/3 / PB
P3

0.2608695652173913

4) В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию. Какова вероятность, что он учится: a). на факультете A б). на факультете B в). на факультете C?

Задача решена вручную с использованием визуальных методик  (см. приложенный файл Excel)

Если бы группы было 2, можно было бы легко решить задачу с помощью теоремы Байеса - можно и с помощью неее вобще-то

Решение с помощью теоремы Байеса
 P(A|B) = P(B|A) * P(A)/P(B)

In [ ]:
# Полноя вероятность события B что студент сдал сессию
PB = 0.25*0.8 + 0.25 * 0.7 + 0.5*0.9 
PB

0.825

In [ ]:
# 3.1) Студент, сдавгий сессию учится в группе A
P1 = 0.8 * 0.25 / PB
P1

0.24242424242424246

In [ ]:
# 3.2) Студент, сдавгий сессию учится в группе B
P2 = 0.7 * 0.25 / PB
P2

0.21212121212121213

In [ ]:
# 3.3) Студент, сдавгий сессию учится в группе C
P3 = 0.9 * 0.5 / PB
P3

0.5454545454545455

5) Устройство состоит из трех деталей. Для первой детали вероятность выйти из строя в первый месяц равна 0.1, для второй - 0.2, для третьей - 0.25. Какова вероятность того, что в первый месяц выйдут из строя: а). все детали б). только две детали в). хотя бы одна деталь г). от одной до двух деталей?

а). все детали

In [ ]:
P_3 = 0.1 * 0.2 * 0.25
P_3

0.005000000000000001

Ответ: все детали могут выйти из строя с вероятностью 0,5%

б). только две детали

In [ ]:
P_2 = 0.1 * 0.2 * (1 - 0.25) + 0.1 * (1 - 0.2) * 0.25 + (1 - 0.1) * 0.2 * 0.25
P_2

0.08000000000000002

Ответ: только 2 детали могут выйти из строя с вероятностью 8%

в). хотя бы одна деталь

Для начала посчитаем вероятность, что ни одна деталь не выйдет и строя


In [ ]:
P_0 = (1 - 0.1) * (1 - 0.2) * (1 - 0.25)
P_0

0.54

In [ ]:
P_1_to_3 = 1 - P_0
P_1_to_3

0.45999999999999996

Ответ: хотя бы одна деталь может выйти из строя с вероятностью 46%

г). от одной до двух деталей

In [ ]:
P_1_to_2 = P_1_to_3 - P_3
P_1_to_2

0.45499999999999996

Ответ: от одной до двух деталей могут выйти из строя с вероятностью 45,5%

Проверка

In [ ]:
P_1 = 0.1 * (1 - 0.2) * (1 - 0.25) + (1 - 0.1) * (1 - 0.2) * 0.25 + (1 - 0.1) * 0.2 * (1 - 0.25)
P_1

0.37500000000000006

Только 1 деталь может выйти из строя с вероятностью 37,5%


Проверка

In [ ]:
P_full = P_0 + P_1 + P_2 + P_3 
P_full

1.0